includes/sanitizers/class-amp-style-sanitizer.php

```diff
@@ -2661,23 +2703,41 @@ private function process_font_face_at_rule( AtRuleSet $ruleset, $options ) {
 					'genericons.woff',
 				];
 				if ( in_array( $font_filename, $bundled_fonts, true ) ) {
-					$data_url->getURL()->setString( plugin_dir_url( AMP__FILE__ ) . "assets/fonts/$font_filename" );
-					$converted_count++;
+					$font_file = plugin_dir_url( AMP__FILE__ ) . "assets/fonts/$font_filename";
+					$data_url->getURL()->setString( $font_file );
+					if ( 'inline' === $first_src_type ) {
+						$first_src_type = 'file';
+						$font_file      = $font_file;
+					}
 				}
 			} // End foreach $source_data_url_objects.
 		} // End foreach $src_properties.
 
-		/*
-		 * If a data: URL has been replaced with an external file URL, then we add a font-display:swap to the @font-face
-		 * rule if one isn't already present. This prevents FO
-		 *
-		 *  If no font-display is already present, add font-display:swap since the font is now being loaded externally.
-		 */
-		if ( $converted_count && 0 === count( $ruleset->getRules( 'font-display' ) ) ) {
+		// Override the 'font-display' property to improve font performance.
+		if ( $font_family && in_array( $font_family, array_keys( $this->args['font_face_display_overrides'] ), true ) ) {
+			$ruleset->removeRule( 'font-display' );
 			$font_display_rule = new Rule( 'font-display' );
-			$font_display_rule->setValue( 'swap' );
+			$font_display_rule->setValue( $this->args['font_face_display_overrides'][ $font_family ] );
 			$ruleset->addRule( $font_display_rule );
 		}
+
+		// If the font-display is auto, block, or swap then we should automatically add the preload link for the first font file.
+		$properties = $ruleset->getRules( 'font-display' );
+		$property   = end( $properties ); // Last since the last property wins in CSS.
+		if (
+			(
+				// Defaults to 'auto', hence should be preloaded as well.
+				! $property instanceof Rule
+				||
+				in_array( $property->getValue(), [ 'auto', 'block', 'swap' ], true )
+			)
+			&&
+			'file' === $first_src_type
+			&&
+			! empty( $font_file )
+		) {
+			$this->dom->links->addPreload( $font_file, RequestDestination::FONT );
```

Comment:
I just realized this can't be done bere because when a parsed stylesheet is cached, this logic won't run. This opens a can of worms, as it means we have to collect the font URLs to preload and then pass them back all the way up to where we store the cached stylesheet, and then we need to add the preloading when we add the processed stylesheet to the document.

Reply:
When I activated the Twenty Fifteen theme, for example, I'd only see the Genericons preload link the first time I visited a page. After reloading it would go away.

Failing test case added in 95c7da2384cb6719625f8755929c8b56121defe0 and fix implemented in b3b6579c5d2cf97b7e885ac98f804d57df55f519.